pkg/providers/amifamily/ami.go

```diff
@@ -54,18 +55,37 @@ type DefaultProvider struct {
 }
 
 type AMI struct {
-	Name         string
-	AmiID        string
-	CreationDate string
-	Requirements scheduling.Requirements
+	Name            string
+	AmiID           string
+	CreationDate    string
+	DeprecationTime string
+	Requirements    scheduling.Requirements
 }
 
 type AMIs []AMI
 
 // Sort orders the AMIs by creation date in descending order.
-// If creation date is nil or two AMIs have the same creation date, the AMIs will be sorted by ID, which is guaranteed to be unique, in ascending order.
+// If there are any amis with deprecation time, they will be sorted by deprecation time in descending order
+// If there deprecation time is nil or two AMIs have the same deprecation time, the AMIs will be sorted by ID, which is guaranteed to be unique, in ascending order.
+// finally, if creation date is nil or two AMIs have the same creation date, the AMIs will be sorted by ID, which is guaranteed to be unique, in ascending order.
 func (a AMIs) Sort() {
+	maxDepTime := time.Unix(math.MaxInt64, 0)
 	sort.Slice(a, func(i, j int) bool {
+		ideptime := lo.TernaryF(
+			a[i].DeprecationTime != "",
+			func() time.Time { t := lo.Must(time.Parse(time.RFC3339, a[i].DeprecationTime)); return t },
+			func() time.Time { return maxDepTime },
+		)
+		jdeptime := lo.TernaryF(
+			a[j].DeprecationTime != "",
+			func() time.Time { t := lo.Must(time.Parse(time.RFC3339, a[j].DeprecationTime)); return t },
+			func() time.Time { return maxDepTime },
+		)
+
+		if ideptime.Unix() != jdeptime.Unix() {
+			return ideptime.Unix() > jdeptime.Unix()
+		}
+
 		itime, _ := time.Parse(time.RFC3339, a[i].CreationDate)
 		jtime, _ := time.Parse(time.RFC3339, a[j].CreationDate)
```

Comment:
Just noticed that we want to do a Must here, too.

Reply:
a lot of unrelated tests with the changes in PR fail when using Must here

```
Summarizing 6 Failures:
  [PANICKED!] NodeClass InstanceProfile Status Controller [It] should resolve the specified instance profile into the status when using instanceProfile field
  /Users/shabbirkagalwala/go/pkg/mod/github.com/samber/lo@v1.44.0/errors.go:53
  [PANICKED!] NodeClass InstanceProfile Status Controller [It] should not call the the IAM API when specifying an instance profile
  /Users/shabbirkagalwala/go/pkg/mod/github.com/samber/lo@v1.44.0/errors.go:53
  [PANICKED!] NodeClass InstanceProfile Status Controller [It] should not call CreateInstanceProfile or AddRoleToInstanceProfile when instance profile exists with correct role
  /Users/shabbirkagalwala/go/pkg/mod/github.com/samber/lo@v1.44.0/errors.go:53
  [PANICKED!] NodeClass InstanceProfile Status Controller [It] should update the role for the instance profile when the wrong role exists
  /Users/shabbirkagalwala/go/pkg/mod/github.com/samber/lo@v1.44.0/errors.go:53
  [PANICKED!] NodeClass InstanceProfile Status Controller [It] should create the instance profile when it doesn't exist
  /Users/shabbirkagalwala/go/pkg/mod/github.com/samber/lo@v1.44.0/errors.go:53
  [PANICKED!] NodeClass InstanceProfile Status Controller [It] should add the role to the instance profile when it exists without a role
  /Users/shabbirkagalwala/go/pkg/mod/github.com/samber/lo@v1.44.0/errors.go:53
```

Reply:
Interesting. We can address separately.

Reply:
updated based on conversation.